blog/content/post/2025-07-xx-edk2-on-odroid-m2.md

```diff
@@ -0,0 +1,325 @@
+standard boot firmware in the x86 space. In the ARM world, especially in the
+embedded Linux space, that role is typically served by U-boot. With the rise of
+high performance ARM computing, however, EDK II is becoming an increasingly
+popular option for system builders. In this blog post I'll be checking out the
+process of booting EDK II and an UEFI OS on an Arm platform with pre-existing
+SoC support.
+
+## The platform
+
+![img](/img/odroid_m2_board.jpg)
+
+I picked an Odroid M2, which is an SBC with a Rockchip RK3588 SoC. I wanted
+that SoC in particular, because there is already some open-source EDK II support
+for it, in the form of [edk2-porting/edk2-rk3588](https://github.com/edk2-porting/edk2-rk3588).
+
+The M2 is a pretty neat little board that can run Ubuntu 25.04 smoothly, has an
+NPU for users who want it, and most of the Linux support code is already merged.
+It also has a 40-pin Raspberry Pi-like GPIO header as well as an extra 2x7 GPIO
+header with extra features. The SBC also features a HDMI port, USB-C with
+DisplayPort, a debug serial console port, HDMI and MIPI DSI, and a gigabit
+Ethernet port.
+
+## The port
+
+I started off by cloning the edk2-rk3588 repo and taking a look at the mainboard
+specific code. To make my job easier, I took the existing OrangePi 5 port and
+started modifying it for my purposes. These were the most important parts I
+needed to modify:
+
+### Voltage regulators
+
+The first thing I would set were the voltage regulator parameters, as I figured
+these would be the most important for safe functioning of the board.
+
+This is the voltage setting code in the OrangePI 5 board code:
+
+```c
+static struct regulator_init_data  rk806_init_data[] = {
+  /* Master PMIC */
+  RK8XX_VOLTAGE_INIT (MASTER_BUCK1,  850000),
+  RK8XX_VOLTAGE_INIT (MASTER_BUCK3,  750000),
+  RK8XX_VOLTAGE_INIT (MASTER_BUCK4,  750000),
+  RK8XX_VOLTAGE_INIT (MASTER_BUCK5,  850000),
+
+  /* This is not configured in the OrangePi5's Linux device tree
+  RK8XX_VOLTAGE_INIT(MASTER_BUCK6, 1100000), */
+  RK8XX_VOLTAGE_INIT (MASTER_BUCK7,  2000000),
+  RK8XX_VOLTAGE_INIT (MASTER_BUCK8,  3300000),
+  RK8XX_VOLTAGE_INIT (MASTER_BUCK10, 1800000),
+
+  RK8XX_VOLTAGE_INIT (MASTER_NLDO1,  750000),
+  RK8XX_VOLTAGE_INIT (MASTER_NLDO2,  850000),
+  /* The OPi is officially configured for the 837500 voltage, but is still marked as avdd_0v75_s0 in the schematic and Linux device tree. rockchip says this voltage is set to improve HDMI stability. */
+  RK8XX_VOLTAGE_INIT (MASTER_NLDO3,  837500),
+  RK8XX_VOLTAGE_INIT (MASTER_NLDO4,  850000),
+  RK8XX_VOLTAGE_INIT (MASTER_NLDO5,  750000),
+
+  RK8XX_VOLTAGE_INIT (MASTER_PLDO1,  1800000),
+  RK8XX_VOLTAGE_INIT (MASTER_PLDO2,  1800000),
+  RK8XX_VOLTAGE_INIT (MASTER_PLDO3,  1200000),
+  RK8XX_VOLTAGE_INIT (MASTER_PLDO4,  3300000),
+  RK8XX_VOLTAGE_INIT (MASTER_PLDO5,  3300000),
+  RK8XX_VOLTAGE_INIT (MASTER_PLDO6,  1800000),
+  /* No dual PMICs on this platform */
+};
+```
+
+vdd_0v75_s3: nldo-reg1 {
```

Comment:
missing code block?

Reply:
https://github.com/3mdeb/news-and-ideas/pull/674/commits/9a49a89fe5a550ac462bc33c79a65fca4533e9bf